### PR TITLE
Fixed list of commands in help and added README documentation on commands and interactivity

### PR DESCRIPTION
**Pull request by olehermanse**
The help menu was looking for functions ending with _command
and taking the first word of those functions. This no longer works
correctly, some commands are multiple words and some of those
functions are not commands.

### CODE REVIEW

**Review by craigcomstock — APPROVED** [review]
looks good to me! thanks!

File: README.md [hunk 1]
```diff
@@ -220,6 +220,76 @@ from pydantic import (
...
+The file contents themselves are (should be) identical.
+There is also no testing of reproducible builds or checking to enforce that what you built locally and what your hub built are the same.
+We'd like to improve all of this, to see the progress in this area, take a look at this JIRA ticket:
+
+https://tracker.mender.io/browse/CFE-4102
+
+## Available commands
+
+We expect policy writers and module authors to use the `cfbs` tooling from the command line, when working on projects, in combination with their editor of choice, much in a similar way to `git`, `cargo`, or `npm`.
+When humans are running the tool in a shell, manually typing in commands, we want it to be intuitive, and have helpful prompts to make it easy to use.
```
Comment (larsewi) [inline README.md:238]:
```suggestion
When humans are running the tool in a shell - manually typing in commands - we want it to be intuitive, and have helpful prompts to make it easy to use.
```

File: README.md [hunk 2]
```diff
@@ -220,6 +220,76 @@ from pydantic import (
...
+  Should work offline if things are already downloaded (by `cfbs download`).
+* `cfbs get-input`: Get input data for a module.
+  Includes both the specification for what the module accepts as well as the user's responses.
+  Can be used on modules not yet added to project to get just the specification.
+* `cfbs install`: Run this on a hub as root to install the policy set (copy the files from `out/masterfiles` to `/var/cfengine/masterfiles`).
+* `cfbs pretty`: Run on a JSON file to pretty-format it. (May be expanded to other formats in the future).
+* `cfbs set-input`: Set input data for a module.
+  Non-interactive version of `cfbs input`, takes the input as a JSON, validates it and stores it.
+  `cfbs set-input` and `cfbs get-input` can be though of as ways to save and load the input file.
+  Similar to `cfbs get-input` the JSON contains both the specification (what the module accepts and how it's presented to the user) as well as the user's responses.
```
Comment (larsewi) [inline README.md:286]:
```suggestion
  Similar to `cfbs get-input` the JSON contains both the specification (what the module accepts and how it's presented to the user) as well as the user's responses (if present).
```

File: tests/shell/common.sh [hunk 3]
```diff
@@ -0,0 +1,13 @@
+git-must-track () {
+  # $1: Path / filename
+  # Error if the file has never been added:
+  git ls-files --error-unmatch $1
+}
```
Comment (larsewi) [inline tests/shell/common.sh:5]: Accepts only one argument?

**Review by larsewi — COMMENTED** [review]
> @vpodzime @larsewi What do you think of this commit specifically?
> https://github.com/cfengine/cfbs/commit/9e87fed3983dab89718fb70c6897bf06746050c2
> I can't really make up my mind what is better 🤷

IDK, maybe the decorator can take a name or tuple of names? Anyways, it will not be needed anymore when we create the "larser" parser supporting sub-parsers and floating global options 😉

File: README.md [hunk 4]
```diff
@@ -220,6 +220,76 @@ from pydantic import (
 $
 ```
 
+## Reproducible builds
+
+When you run `cfbs build` locally, and when your hub runs it after pulling the latest changes from git, the resulting policy sets should be identical.
+(This does not extend to providing reproducibility after running arbitrary combinations of `cfbs init`, `cfbs add` commands, etc.).
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply README.md:226]: If I run `cfbs init && cfbs add masterfiles && cfbs build` today and you run it on your computer tomorrow, we don't say that this should be reproducible. Versions of modules change and behavior of `init` and `add` can change.

File: tests/shell/common.sh (hunk 3 above)
Comment (olehermanse) [reply tests/shell/common.sh:5]: Yep, I don't overengineer for something I don't need yet :)

File: README.md [hunk 5]
```diff
@@ -220,6 +220,76 @@ from pydantic import (
 $
 ```
 
+## Reproducible builds
+
+When you run `cfbs build` locally, and when your hub runs it after pulling the latest changes from git, the resulting policy sets should be identical.
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply README.md:225]: When I run **X**, and when you run **it** (referring to the same thing) is quite a common construction, I think it reads okay.

File: README.md (hunk 1 above)
Comment (olehermanse) [reply README.md:238]: NACK

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (olehermanse) [reply README.md:226]: We want `cfbs build` to be stable, to provide the same result every time. We don't want `cfbs init` or `cfbs add` to have this same expectation.

**Review by larsewi — APPROVED** [review]

File: cfbs/commands.py [hunk 6]
```diff
@@ -64,6 +65,39 @@ def __init__(self, message):
 FIRST_ARG_SLIST = lambda args, _: ", ".join("'%s'" % module for module in args[0])
 
 
+def get_command_names():
+    # TODO: Consider making a decorator for command functions to make this
+    #       less error prone.
```
Comment (vpodzime) [inline cfbs/commands.py]: :+1:

File: cfbs/commands.py [hunk 7]
```diff
@@ -64,6 +65,39 @@ def __init__(self, message):
...
+        "update",
+        "validate",
+    )
+    defined = dir(sys.modules[__name__])
+    for command in names:
+        function_name = command.replace("-", "_") + "_command"
+        assert function_name in defined, "'%s' function not found in %s" % (
+            function_name,
+            defined,
+        )
```
Comment (vpodzime) [inline cfbs/commands.py]: Do we really need this check?

File: README.md (hunk 5 above)
Comment (vpodzime) [inline README.md:225]: *will be identical*

File: cfbs/commands.py [hunk 8]
```diff
@@ -64,30 +64,23 @@ def __init__(self, message):
...
-        "pretty",
-        "remove",
-        "search",
-        "set-input",
-        "show",
-        "status",
-        "update",
-        "validate",
-    )
+    global _commands
```
Comment (vpodzime) [inline cfbs/commands.py:83]: Same here.

File: cfbs/commands.py [hunk 9]
```diff
@@ -20,6 +20,7 @@
     user_error,
     strip_right,
     pad_right,
+    ProgrammerError,
```
Comment (vpodzime) [inline cfbs/commands.py:23]: Where is the new exception being added?

File: cfbs/commands.py [hunk 10]
```diff
@@ -965,6 +958,7 @@ def _print_module_info(data):
             print("{}: {}".format(key.title().replace("_", " "), value))
 
 
+@cfbs_command("show")
```
Comment (vpodzime) [inline cfbs/commands.py:961]: I'd squash this commit with the one introducing the decorator.

**Review by vpodzime — COMMENTED** [review]
Generally I think the commits should be squashed to leave out the intermediate wrong/sub-optimal states.

File: cfbs/commands.py (hunk 7 above)
Comment (olehermanse) [reply cfbs/commands.py]: It was very helpful while I was refactoring, but we can remove it now (I already did in a later commit, will squash).

File: cfbs/commands.py (hunk 9 above)
Comment (olehermanse) [reply cfbs/commands.py:23]: It was added in the first commit of this PR, but I can rewrite history to make it more clear :)

File: cfbs/commands.py (hunk 10 above)
Comment (olehermanse) [reply cfbs/commands.py:961]: Will do!

File: README.md (hunk 5 above)
Comment (olehermanse) [reply README.md]: Thanks, added.

**Pull request merged**